# ASSIGNMENT - 4
# NAME :ARVIND SHANKAR RAMAN
# NET-ID : AXS170059

## QUESTION:
### BUAN6340 Programming for Data Science (Python)
### 1.	Please download spam.csv from eLearning.
### 2.	Train a Naïve Bayes model to predict spam.
### 3.	Report the accuracy rate for your training and test datasets.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [3]:
import nltk

In [4]:
import pandas as pd
spam = pd.read_csv('spam.csv', encoding='ISO-8859-1');

In [5]:
spam.drop(spam.columns[2:5], axis=1,inplace = True)

In [6]:
spam.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
v1    5572 non-null object
v2    5572 non-null object
dtypes: object(2)
memory usage: 87.1+ KB


## Tokenizing the spam v2 column

In [7]:
spam['v2']=spam['v2'].apply(nltk.word_tokenize)

### Putting the tokenized words in to the bag of words

In [8]:
def build_bag_of_words_features(words):
    return {word:True for word in words}

In [9]:
spam['v2'] = spam.v2.apply(lambda x : (build_bag_of_words_features(x)))

### Removing stopwords and creating a filtered bag of words

In [10]:
nltk.download("stopwords")

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\91909\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [11]:
nltk.corpus.stopwords.words("english")

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [12]:
from nltk.stem.porter import PorterStemmer
stemmer = PorterStemmer()

In [13]:
spam['v2'] = spam.v2.apply(lambda row : [stemmer.stem(word) for word in row])

In [14]:
spam['v2'][0]

['Go',
 'until',
 'jurong',
 'point',
 ',',
 'crazy..',
 'avail',
 'onli',
 'in',
 'bugi',
 'n',
 'great',
 'world',
 'la',
 'e',
 'buffet',
 '...',
 'cine',
 'there',
 'got',
 'amor',
 'wat']

In [15]:
import string

In [16]:
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [17]:
useless_words = nltk.corpus.stopwords.words("english") + list(string.punctuation)

In [18]:
def build_bag_of_words_features_filtered(words):
    return {word:1 for word in words if not word in useless_words}

In [19]:
spam['v2'] = spam.v2.apply(lambda x : (build_bag_of_words_features_filtered(x)))

In [20]:
spam.v2[0]

{'Go': 1,
 'jurong': 1,
 'point': 1,
 'crazy..': 1,
 'avail': 1,
 'onli': 1,
 'bugi': 1,
 'n': 1,
 'great': 1,
 'world': 1,
 'la': 1,
 'e': 1,
 'buffet': 1,
 '...': 1,
 'cine': 1,
 'got': 1,
 'amor': 1,
 'wat': 1}

### Separating the filtered column into positive features and negative features.

In [21]:
for i in range(len(spam)):
    if(spam['v1'][i] == 'ham'):
        spam['v2'][i]=(spam['v2'][i],'ham')
    if(spam['v1'][i] == 'spam'):
        spam['v2'][i]=(spam['v2'][i],'spam')

In [22]:
positive_features=spam['v2'][spam['v1']=='ham']

In [23]:
negative_features=spam['v2'][spam['v1']=='spam']

In [24]:
len(positive_features)
type(positive_features)

4825

pandas.core.series.Series

In [25]:
len(negative_features)
type(negative_features)

747

pandas.core.series.Series

In [26]:
pd.concat([positive_features, negative_features])

0       ({'Go': 1, 'jurong': 1, 'point': 1, 'crazy..':...
1       ({'Ok': 1, 'lar': 1, '...': 1, 'joke': 1, 'wif...
3       ({'U': 1, 'dun': 1, 'say': 1, 'earli': 1, 'hor...
4       ({'nah': 1, 'I': 1, 'n't': 1, 'think': 1, 'goe...
6       ({'even': 1, 'brother': 1, 'like': 1, 'speak':...
7       ({'As': 1, 'per': 1, 'request': 1, ''mell': 1,...
10      ({'I': 1, ''m': 1, 'gon': 1, 'na': 1, 'home': ...
13      ({'I': 1, ''ve': 1, 'search': 1, 'right': 1, '...
14      ({'I': 1, 'A': 1, 'date': 1, 'ON': 1, 'sunday'...
16      ({'Oh': 1, 'k': 1, '...': 1, ''m': 1, 'watch':...
17      ({'Eh': 1, 'u': 1, 'rememb': 1, '2': 1, 'spell...
18      ({'fine': 1, 'thatåõ': 1, 'way': 1, 'u': 1, 'f...
20      ({'Is': 1, 'serious': 1, 'spell': 1, 'hi': 1, ...
21      ({'i‰û÷m': 1, 'go': 1, 'tri': 1, '2': 1, 'mont...
22      ({'So': 1, 'Ì_': 1, 'pay': 1, 'first': 1, 'lar...
23      ({'aft': 1, 'finish': 1, 'lunch': 1, 'go': 1, ...
24      ({'ffffffffff': 1, 'alright': 1, 'way': 1, 'I'...
25      ({'for

### Runnning the Naive Bayes Classifier Model on the positive feautres

In [27]:
from nltk.classify import NaiveBayesClassifier

In [28]:
split = int(0.8*len(spam))
split

4457

### Training the model

In [29]:
sentiment_classifier = NaiveBayesClassifier.train(pd.concat([positive_features[:split], negative_features[:split]]))

In [30]:
print(positive_features[0])

({'Go': 1, 'jurong': 1, 'point': 1, 'crazy..': 1, 'avail': 1, 'onli': 1, 'bugi': 1, 'n': 1, 'great': 1, 'world': 1, 'la': 1, 'e': 1, 'buffet': 1, '...': 1, 'cine': 1, 'got': 1, 'amor': 1, 'wat': 1}, 'ham')


### Calculating the training accuracy rate.

In [31]:
nltk.classify.util.accuracy(sentiment_classifier, pd.concat([positive_features[:split], negative_features[:split]]))*100

94.48501152959263

### Calculating the testing accuracy rate.

In [32]:
nltk.classify.util.accuracy(sentiment_classifier, pd.concat([positive_features[split:], negative_features[split:]]))*100

92.3913043478261

#### The training accuracy rate obtained  is 94.48
#### The testing accuracy rate obtained is  92.39

In [33]:
sentiment_classifier.show_most_informative_features()

Most Informative Features
                   award = 1                spam : ham    =    228.5 : 1.0
                 voucher = 1                spam : ham    =    141.1 : 1.0
                    code = 1                spam : ham    =    117.2 : 1.0
                   await = 1                spam : ham    =    109.3 : 1.0
                  urgent = 1                spam : ham    =    104.7 : 1.0
                      16 = 1                spam : ham    =     97.3 : 1.0
                   nokia = 1                spam : ham    =     89.4 : 1.0
                 landlin = 1                spam : ham    =     82.2 : 1.0
                     txt = 1                spam : ham    =     82.1 : 1.0
                deliveri = 1                spam : ham    =     81.5 : 1.0


# Conclusion:
### 1. The spam.csv was downloaded and the Naive Bayes model was trained to predict spam.
### 2. The training accuracy rate obtained  is 94.48 and the testing accuracy rate obtained is  92.39.
### 3. Also, the top 3 most_informative_features were award, voucher and code.